KAFKA-16093: Fix spurious REST-related warnings on Connect startup

C0urante opened the pull request:
[Jira](https://issues.apache.org/jira/browse/KAFKA-16093)

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

C0urante committed:
MINOR: Remove empty @Path annotations for Connect REST server, fixing spurious warnings

C0urante committed:
MINOR: Implement ContainerRequestFilter interface with no-op in Connect REST resources, fixing spurious warnings

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectResource.java:
```diff
@@ -16,12 +16,15 @@
  * This interface defines shared logic for all Connect REST resources.
  */
-public interface ConnectResource {
+public interface ConnectResource extends ContainerRequestFilter {
```
I found an alternative workaround, it's up to you if you like it better or worse than adding the no-op filter method.

The logic for emitting that warning in `Providers.checkProviderRuntime` is checking that there is something to constrain the component being checked to the current runtime (client vs server). The ContainerRequestFilter is marked as introducing a SERVER constraint in `Providers.getJaxRsProviderInterfaces`, and by extending it we are just inheriting that constraint.

Instead we can use the annotation mechanism intended for adding constraints to third-party SPIs to make the ConnectResource itself look like an SPI to Jersey. I tried adding this to the classes themselves next to the `Singleton` annotations, and it exploded. But it seems to work when the annotations are here on the interface instead.

```suggestion
@Contract
@ConstrainedTo(RuntimeType.SERVER)
public interface ConnectResource {
```

C0urante replied:
They're both hacks, and both require implementing classes to be annotated with `@Singleton` to avoid unnecessary warnings, but this seems slightly cleaner (especially since it doesn't automatically implement the request filter interface that we may want to use for real later on).

I did some local experimenting and it seems like the `@ConstrainedTo` annotation isn't actually necessary--let me know if I'm missing something, though.

gharris1727 replied:
> I did some local experimenting and it seems like the @ConstrainedTo annotation isn't actually necessary--let me know if I'm missing something, though.

I think you're right! I think that was from a slightly earlier iteration I had where i was trying to target [this first condition](https://github.com/eclipse-ee4j/jersey/blob/74bf3220a2dd5f2067e60c110d4981591e87604e/core-common/src/main/java/org/glassfish/jersey/internal/inject/Providers.java#L439) until I realized I couldn't get that one to work, and started targeting [this second condition](https://github.com/eclipse-ee4j/jersey/blob/74bf3220a2dd5f2067e60c110d4981591e87604e/core-common/src/main/java/org/glassfish/jersey/internal/inject/Providers.java#L443) instead.

gharris1727 reviewed (commented):
Oh my... I thought these warnings were here to stay! I've seen them so many times that I couldn't understand what they meant, even though they are very direct about what is incorrect.

Thank you so much for fixing this!

C0urante committed:
Change to @Contract annotation instead of extending ContainerRequestFilter

gharris1727 reviewed (approved):
LGTM, thanks Chris! No more nuisance stderr from tests!

C0urante committed:
Use AbstractBinder and @Inject annotations to correctly register Jersey resources

C0urante commented:
Thanks @gharris1727!

I wanted to take a stab at fixing these errors "correctly", i.e., by registering classes instead of instances with the regular and admin `ResourceConfig`. I've pushed a commit that does this; it's a bit lengthier, but it does work. Let me know what you think; happy to revert and stick with the hack if we'd rather minimize the diff.

C0urante commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/RestServer.java:
```diff
@@ -226,20 +235,20 @@ protected final void initializeResources() {
+            Collection<Class<?>> adminResources = adminResources();
             adminResources.forEach(adminResourceConfig::register);
             adminResourceConfig.register(ConnectExceptionMapper.class);
+            adminResourceConfig.property(ServerProperties.WADL_FEATURE_DISABLE, true);
```
This fixes another spurious warning message that gets emitted if the worker is configured with the `admin.listeners` property.

gharris1727 replied:
this is some light duplication, perhaps the ResourceConfig constructor and the 4 lines accompanying it (request timeout, jackson, exception mapper, and wadl feature disable) can be deduplicated.

C0urante replied:
Good call, done. I've also cleaned up the configuration of the `adminResourceConfig` to hopefully prevent other kinds of duplication-related issues in the future; LMKWYT

gharris1727 commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResourceTest.java:
```diff
@@ -159,7 +163,8 @@ public class ConnectorsResourceTest {
         when(serverConfig.topicTrackingEnabled()).thenReturn(true);
         when(serverConfig.topicTrackingResetEnabled()).thenReturn(true);
-        connectorsResource = new ConnectorsResource(herder, serverConfig, restClient);
+        RestRequestTimeout requestTimeout = () -> RestServer.DEFAULT_REST_REQUEST_TIMEOUT_MS;
```
nit: unused

C0urante replied:
Good catch, thanks 👍

gharris1727 reviewed (commented):
I really like this approach, and I agree it's the non-hacky way forward. TIL about AbstractBinder and the dependency injection mechanism that Jersey provides.

In particular I like removing ConnectResource, and making the RestRequestTimeout object mutable instead of having all of the resources themselves mutable.

C0urante committed:
Address review comments

- Deduplicate RestServer instantiation and configuration of ResourceConfig instances
- Remove unused RestRequestTimeout instance from ConnectorsResourceTest
- Fix failing ConnectRestServerTest suite

gharris1727 reviewed (approved):
LGTM, thanks Chris!

vamossagar12 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/HerderRequestHandler.java:
```diff
@@ -41,18 +41,11 @@ public class HerderRequestHandler {
     private final RestClient restClient;
 
-    private volatile long requestTimeoutMs;
+    private final RestRequestTimeout requestTimeout;
```
The field `requestTimeoutMs` was made volatile as part of [here](https://github.com/apache/kafka/pull/14562/files#diff-f2311f0c356f882d7768b97cb5f0054dc7040d29d455eab74ab585725454488aR44). I don't think we need it now, but just wanted to understand why was it added over there (you can skip the explanation if it's too complex :) )

C0urante replied:
This field is final, it wouldn't make sense to mark it volatile.

The underlying mutable value that it returns from `timeoutMs()` is still marked volatile in the only non-test implementation [here](https://github.com/C0urante/kafka/blob/0713736b9884084c360126558e4b6d92d4435bfa/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/RestServer.java#L533).

vamossagar12 replied:
I didn't intend to mark the final field as volatile. I wanted to know why the field was marked as volatile in the old PR since you had made the change.

C0urante replied:
Ah, gotcha. The intent was to make it thread-safe since it's likely that writes and reads to/of that value will take place on different threads.

vamossagar12 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/ConnectRestServer.java:
```diff
@@ -45,25 +45,40 @@ public void initializeResources(Herder herder) {
+        resourceConfig.register(new Binder());
+    }
+
+    private class Binder extends AbstractBinder {
+        @Override
+        protected void configure() {
+            bind(herder).to(Herder.class);
+            bind(restClient).to(RestClient.class);
+            bind(config).to(RestServerConfig.class);
+        }
+    }
+
```
nit: We could probably move the class definition to the bottom so that all `configureXXXResources` method are together

C0urante replied:
This is true, but not a blocker.

vamossagar12 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/HerderRequestHandler.java:
```diff
@@ -41,18 +41,11 @@ public class HerderRequestHandler {
-    }
-
-    public void requestTimeoutMs(long requestTimeoutMs) {
-        if (requestTimeoutMs < 1) {
-            throw new IllegalArgumentException("REST request timeout must be positive");
-        }
```
I am not sure if this validation was needed in the past as well, but now I don't see it being present. Do you think it's needed?

C0urante replied:
I removed it for two reasons:

1. This code path is only ever used in testing code
2. It's unclear that the validation would be safer than allowing negative values. At least on my machine, negative values don't cause any issues with, e.g., `ConvertingFutureCallback::get`.

vamossagar12 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/RestRequestTimeout.java:
```diff
@@ -0,0 +1,26 @@
+ */
+package org.apache.kafka.connect.runtime.rest;
+
+public interface RestRequestTimeout {
```
nit: Mark this interface as `@FunctionalInterface`?

C0urante replied:
I could see some benefit to this, but it's also not a blocker.

It's also debatable, since this interface may change in the future and we don't want to imply that it can only have exactly one abstract method.

vamossagar12 replied:
Ok. Makes sense.

vamossagar12 reviewed (changes requested):
Thanks for looking into these log messages! Added some nit level comments.

C0urante commented:
Thanks @gharris1727!

@vamossagar12 I appreciate the review but none of these seem like blocking comments. Please try to use the "request changes" button sparingly. Thanks!

Test failures appear unrelated and all come from known flaky tests. Merging...

vamossagar12 commented:
@C0urante , hmm okay. I understand those weren't blocker comments and I called them as nits.
